### PR TITLE
Fix failure store pipeline-level failure recording issues

### PR DESCRIPTION
**Pull request by jbaiera**
This PR does a couple of things:

First, it adds more tests to the failure store redirect yaml test suite. These tests ensure that failures that occur after changing the document's index are correctly routed to the data stream that the document is being considered for during the ingest pipeline execution, not the index the document would be routed to afterward.

Secondly, this PR fixes a couple of bugs found while adding these additional tests:
* The index being used in the failure document was using the target index instead of the index present on the request. If a document fails during ingest along a chain of reroute processors, this bug meant that the failure document was persisted with the index name set to the data stream whose failure store it was saved under instead of the original index name it was sent to the cluster with.
* Added a new exception type to capture pipeline information to be saved on the failure document in cases where ingest operations fail outside of the compound processor. Now, if a document fails ingest because of an erroneous reroute or from a script processor making an unserializable cyclical reference, we record the name of the pipeline on the exception and also on the failure document.

### CODE REVIEW

**Comment by jbaiera** [comment]
@elasticmachine update branch

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/190_failure_store_redirection.yml [hunk 1]
```diff
@@ -318,4 +318,433 @@ teardown:
...
+          template:
+            settings:
+              number_of_shards: 1
+              number_of_replicas: 1
+              index:
+                default_pipeline: "send_to_destination"
+
+  - do:
+      allowed_warnings:
+        - "index template [generic_logs_template] has index patterns [logs-*] matching patterns from existing older templates [global] with patterns (global => [*]); this template [generic_logs_template] will take precedence during new index creation"
```
Comment (nielsbauman) [inline modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/190_failure_store_redirection.yml:573]: After having forgotten to update this myself a couple of times (resulting in a failing test afterwards), I finally remembered to update the string in advance :D
```suggestion
        - "index template [destination_logs_template] has index patterns [destination-*] matching patterns from existing older templates [global] with patterns (global => [*]); this template [destination_logs_template] will take precedence during new index creation"
```

**Review by nielsbauman — APPROVED** [review]
I caught one outdated warning header string. Other than that, this LGTM! Thanks :)

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/190_failure_store_redirection.yml (hunk 1 above)
Comment (jbaiera) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/190_failure_store_redirection.yml:573]: nice catch!

**Pull request merged**